lib/ansible/utils/collection_loader.py

```diff
@@ -0,0 +1,305 @@
+
+        return None
+
+    @staticmethod
+    def _extend_path_with_ns(path, ns):
+        ns_path_add = ns.rsplit('.', 1)[-1]
+
+        return os.path.join(path, ns_path_add)
+
+    def get_data(self, filename):
+        with open(filename, 'rb') as fd:
+            return fd.read()
+
+
+class AnsibleFlatMapLoader(object):
+    _extension_blacklist = ['.pyc', '.pyo']
+    def __init__(self, root_package):
+        self._root_package = root_package
+        self._dirtree = None
+
+    def _init_dirtree(self):
+        # FIXME: thread safety
+        root_path = os.path.dirname(self._root_package.__file__)
+        flat_files = []
+        # FIXME: make this a dict of filename->dir for faster direct lookup?
+        # FIXME: deal with _ prefixed deprecated files (or require another method for collections?)
+        # FIXME: fix overloaded filenames (eg, rename Windows setup to win_setup)
+        for root, dirs, files in os.walk(root_path):
+            # add all files in this dir that don't have a blacklisted extension
+            flat_files.extend(((root, f) for f in files if not any((f.endswith(ext) for ext in self._extension_blacklist))))
+        self._dirtree = flat_files
+
+    def find_file(self, filename):
+        try:
+            # FIXME: thread safety
+            if not self._dirtree:
+                self._init_dirtree()
+
+            if '.' not in filename:  # no extension specified, use extension regex to filter
+                extensionless_re = re.compile(r'^{0}(\..+)?$'.format(re.escape(filename)))
+                # why doesn't Python have first()?
+                try:
+                    # FIXME: store extensionless in a separate direct lookup?
+                    filepath = next(os.path.join(r, f) for r, f in self._dirtree if extensionless_re.match(f))
+                except StopIteration:
+                    raise IOError("couldn't find {0}".format(filename))
+            else:  # actual filename, just look it up
+                # FIXME: this case sucks; make it a lookup
+                try:
+                    filepath = next(os.path.join(r, f) for r, f in self._dirtree if f == filename)
+                except StopIteration:
+                    raise IOError("couldn't find {0}".format(filename))
+
+            return filepath
+        except Exception as ex:
+            print("bang")
+            raise
+
+    def get_data(self, filename):
+        found_file = self.find_file(filename)
+
+        with open(found_file, 'rb') as fd:
+            data = fd.read()
+
+        return data
+
+
+# TODO: implement these for easier inline debugging?
+#    def get_source(self, fullname):
+#    def get_code(self, fullname):
+#    def is_package(self, fullname):
+
+
+def is_collection_qualified_role_name(candidate_name):
+    """
+    Returns true if a name looks like a collection-qualified name (eg, ns.collection.rolename)
+    """
+    if _collection_role_name_re.match(candidate_name):
```

Comment:
```suggestion
    return bool(_collection_role_name_re.match(candidate_name))
```

Reply:
this function ended up not being used, so just killed it